library/pkparse.c

```diff
@@ -1250,11 +1250,45 @@ int mbedtls_pk_parse_public_key( mbedtls_pk_context *ctx,
 {
     int ret;
     unsigned char *p;
+#if defined(MBEDTLS_RSA_C)
+    const mbedtls_pk_info_t *pk_info;
+#endif
 #if defined(MBEDTLS_PEM_PARSE_C)
     size_t len;
     mbedtls_pem_context pem;
 
     mbedtls_pem_init( &pem );
+#if defined(MBEDTLS_RSA_C)
+    /* Avoid calling mbedtls_pem_read_buffer() on non-null-terminated string */
+    if( keylen == 0 || key[keylen - 1] != '\0' )
+        ret = MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT;
+    else
+        ret = mbedtls_pem_read_buffer( &pem,
+                               "-----BEGIN RSA PUBLIC KEY-----",
+                               "-----END RSA PUBLIC KEY-----",
+                               key, NULL, 0, &len );
+
+    if( ret == 0 )
+    {
+        p = pem.buf;
+        if( ( pk_info = mbedtls_pk_info_from_type( MBEDTLS_PK_RSA ) ) == NULL )
+            return( MBEDTLS_ERR_PK_UNKNOWN_PK_ALG );
+
+        if( ( ret = mbedtls_pk_setup( ctx, pk_info ) ) != 0 )
+            return( ret );
+
+        if ( ( ret = pk_get_rsapubkey( &p, p + pem.buflen, mbedtls_pk_rsa( *ctx ) ) ) != 0 )
+            mbedtls_pk_free( ctx );
+
+        mbedtls_pem_free( &pem );
+        return( ret );
+    }
+    else if( ret != MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT )
+    {
+        mbedtls_pem_free( &pem );
+        return( ret );
+    }
```

Comment:
There is a slight abstraction break happening here (without consequences): If we fall through the RSA PEM case, then `pem` is no longer guaranteed by the API to be in initialized state (it has already been used in `mbedtls_pem_read_buffer` after `mbedtls_pem_init`), and hence there should be a call to `mbedtls_pem_free` and `mbedtls_pem_init` before the next call to `mbedtls_pem_read_buffer`. @RonEld Could you add such a pair at the end of the `MBEDTLS_RSA_C` block?

Reply:
@hanno-arm I may be not following you. This is same logic as in `mbedtls_pk_parse_key` , isn't it?

Reply:
@RonEld The point is that `pem` should be in a freshly initialized state when we fall through to the next call of `mbedtls_pem_read_buffer` a few lines below -- but, at least as far as the API is concerned, that's not guaranteed, because we have already called `mbedtls_pem_read_buffer` once in between. Calling `mbedtls_pem_free( &pem ); mbedtls_pem_init( &pem );` at the end of the `MBEDTLS_RSA_C` block would fix that.

Reply:
@RonEld You're right, the private key parsing functions do it in the same not totally correct, yet non-critical, way. So let's keep it for now.